### PR TITLE
*: refactor kv rpc implementation

### PR DESCRIPTION
**Pull request by xiang90**
This pull request refactors the kv implementation. The main goal is to make the workflow and interface clean. It duplicates a few code for readability purpose.

The main change is to replace the `V3Demo` methods with a few smaller and well-defined methods like put, delete.

### CODE REVIEW

File: etcdmain/etcd.go [hunk 1]
```diff
@@ -321,7 +321,7 @@ func startEtcd(cfg *config) (<-chan struct{}, error) {
 	if cfg.v3demo {
 		// set up v3 demo rpc
 		grpcServer := grpc.NewServer()
-		etcdserverpb.RegisterKVServer(grpcServer, v3rpc.New(s))
+		etcdserverpb.RegisterKVServer(grpcServer, v3rpc.NewKV(s))
```
Comment (yichengq) [inline etcdmain/etcd.go]: NewKVServer? This name sounds natural when reading `NewWatchServer` below, and `v3rpc` has no hint for the server.

File: etcdserver/v3demo_server.go [hunk 2]
```diff
@@ -23,8 +23,52 @@ import (
 	dstorage "github.com/coreos/etcd/storage"
 )
 
-type V3DemoServer interface {
-	V3DemoDo(ctx context.Context, r pb.InternalRaftRequest) (proto.Message, error)
+type RaftKV interface {
```
Comment (yichengq) [inline etcdserver/v3demo_server.go]: add some comment for it? i cannot understand why it is named RaftKV.

Comment (xiang90) [inline etcdserver/v3demo_server.go]: I will add docs when I feel good about the whole thing. I am still in the process of cleaning this up. The main goal of the pr is to separate things up.

Comment (yichengq) [inline etcdserver/v3demo_server.go]: ok

**Comment by yichengq** [comment]
lgtm

**Comment by xiang90** [comment]
@yichengq I will wait for @jonboulle's opinion on this a little bit.

**Comment by xiang90** [comment]
@jonboulle Timeout :) But please write comments. I will address them.

File: etcdserver/v3demo_server.go [hunk 3]
```diff
@@ -32,12 +76,12 @@ type applyResult struct {
 	err  error
 }
 
-func (s *EtcdServer) V3DemoDo(ctx context.Context, r pb.InternalRaftRequest) (proto.Message, error) {
+func (s *EtcdServer) processInternalRaftReq(ctx context.Context, r pb.InternalRaftRequest) (*applyResult, error) {
```
Comment (jonboulle) [inline etcdserver/v3demo_server.go]: `processInternalRaftRequest` - the abbreviation is a bit arbitrary/inconsistent

**Comment by jonboulle** [comment]
Seems fine but like Yicheng I would find it clearer to understand the motivation here with some docstrings.

File: etcdserver/v3demo_server.go (hunk 3 above)
Comment (xiang90) [inline etcdserver/v3demo_server.go]: Will change.

**Pull request merged**